src/system/SystemPoolHeap.h

```diff
@@ -0,0 +1,107 @@
+/*
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    Copyright (c) 2016-2017 Nest Labs, Inc.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <lib/core/CHIPConfig.h>
+
+#if CHIP_SYSTEM_CONFIG_POOL_USE_HEAP
+
+#include <mutex>
+#include <set>
+
+#include <lib/support/CodeUtils.h>
+#include <system/SystemPoolStatistics.h>
+
+namespace chip {
+namespace System {
+
+/**
+ *  @brief
+ *      A class template used for allocating Object subclass objects from an ObjectArena<> template union.
+ *
+ *  @tparam     T   a subclass of Object to be allocated from the arena.
+ *  @tparam     N   a positive integer number of objects of class T to allocate from the arena.
+ */
+template <class T, size_t N>
+class ObjectPoolHeap : public ObjectPoolStatistics
+{
+public:
+    template <typename... Args>
+    T * CreateObject(Args &&... args)
+    {
+        std::lock_guard<std::recursive_mutex> lock(mutex);
```

Comment:
Why are we adding locking here? It should be up to the consumers of this object to enforce the right threading constructs here.

Reply:
I haven't checked all the usage of the pool. The lock may not be required here. But, because this is generically a refactor, the original SystemObject pool is written as thread-safe, so I kept the behavior here.

Reply:
Right.  A memory allocator really needs to be thread-safe or we're going to have an extremely tough time.

We currently have two implementations: one with a mutex (heap) and one that uses atomics (non-heap).  That seemed reasonable to me.

Reply:
We do need lock here to pass multi-thread unit test, but we should not change std::mutex to std::recursive_mutex, I remember we have already discussed if allow recursive locking in our SDK, the consensus is NO since recursive mutexes encourage you to completely lose track of your locking scheme and scope. And hide bug which will be extremely difficult to debug later